packages/core-plugin-api/src/adaptable-components/types.ts

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ComponentType, RefAttributes } from 'react';
+
+import { BackstagePlugin } from '../plugin';
+
+/** @public */
+export type OpaqueComponentProps<Context extends {}> = { value: Context };
+
+/** @public */
+export type ComponentRootProviderProps<Props extends {}, Context> = {
+  /**
+   * An opaque component representing the _inner_ component (or the next
+   * adaptation) with the default context value.
+   */
+  Component: ComponentType<OpaqueComponentProps<Context>>;
```

Comment:
`ComponentType` is kinda meh because it adds `children` regardless of whether the component actually renders them or not. We tend to use something like `(props: TProps) => JSX.Element | null` instead.

Reply:
This is fixed in React 18 where `children` is no longer a default prop.

Reply:
Yep but not there yet. Feeling it's better to keep things aligned for now and then perhaps in the future we can migrate over to using the new `ComponentType`.